pkg/git/remote_git_resolver.go

```diff
@@ -2,59 +2,94 @@ package git
 
 import (
 	gogit "github.com/go-git/go-git/v5"
-	"github.com/go-git/go-git/v5/config"
 	"github.com/go-git/go-git/v5/plumbing"
 	"github.com/go-git/go-git/v5/plumbing/transport"
 	"github.com/go-git/go-git/v5/storage/memory"
-
 	corev1alpha1 "github.com/pivotal/kpack/pkg/apis/core/v1alpha1"
+	"regexp"
+	"strings"
 )
 
-const defaultRemote = "origin"
-
 type remoteGitResolver struct{}
 
 func (*remoteGitResolver) Resolve(auth transport.AuthMethod, sourceConfig corev1alpha1.SourceConfig) (corev1alpha1.ResolvedSourceConfig, error) {
-	remote := gogit.NewRemote(memory.NewStorage(), &config.RemoteConfig{
-		Name: defaultRemote,
-		URLs: []string{sourceConfig.Git.URL},
-	})
+	var resolvedConfig corev1alpha1.ResolvedSourceConfig
 
-	refs, err := remote.List(&gogit.ListOptions{
+	r, err := gogit.Clone(memory.NewStorage(), nil, &gogit.CloneOptions{
+		URL:  sourceConfig.Git.URL,
 		Auth: auth,
 	})
+
 	if err != nil {
-		return corev1alpha1.ResolvedSourceConfig{
+		resolvedConfig = corev1alpha1.ResolvedSourceConfig{
 			Git: &corev1alpha1.ResolvedGitSource{
 				URL:      sourceConfig.Git.URL,
 				Revision: sourceConfig.Git.Revision,
 				Type:     corev1alpha1.Unknown,
 				SubPath:  sourceConfig.SubPath,
 			},
-		}, nil
+		}
 	}
 
-	for _, ref := range refs {
+	rIter, _ := r.References()
+
+	rIter.ForEach(func(ref *plumbing.Reference) error {
 		if ref.Name().Short() == sourceConfig.Git.Revision {
-			return corev1alpha1.ResolvedSourceConfig{
+			refSourceType := sourceType(ref)
+			effectiveCommit := ""
+
+			if refSourceType == corev1alpha1.Branch {
+				effectiveCommit, err = getLogs(r, sourceConfig.SubPath)
+			} else {
+				effectiveCommit = ref.Hash().String()
+			}
+
+			if err != nil {
+				return nil
+			}
+
+			resolvedConfig = corev1alpha1.ResolvedSourceConfig{
 				Git: &corev1alpha1.ResolvedGitSource{
 					URL:      sourceConfig.Git.URL,
-					Revision: ref.Hash().String(),
-					Type:     sourceType(ref),
+					Revision: effectiveCommit,
+					Type:     refSourceType,
 					SubPath:  sourceConfig.SubPath,
 				},
-			}, nil
+			}
+		}
+		return nil
+	})
+
+	if resolvedConfig.ResolvedSource() == nil {
+		resolvedConfig = corev1alpha1.ResolvedSourceConfig{
+			Git: &corev1alpha1.ResolvedGitSource{
+				URL:      sourceConfig.Git.URL,
+				Revision: sourceConfig.Git.Revision,
+				Type:     corev1alpha1.Commit,
+				SubPath:  sourceConfig.SubPath,
+			},
 		}
 	}
 
-	return corev1alpha1.ResolvedSourceConfig{
-		Git: &corev1alpha1.ResolvedGitSource{
-			URL:      sourceConfig.Git.URL,
-			Revision: sourceConfig.Git.Revision,
-			Type:     corev1alpha1.Commit,
-			SubPath:  sourceConfig.SubPath,
+	return resolvedConfig, err
+}
+
+func getLogs(r *gogit.Repository, subPath string) (string, error) {
+	logOutput, err := r.Log(&gogit.LogOptions{
+		PathFilter: func(s string) bool {
+			if strings.Contains(s, subPath) {
+				return true
+			} else {
+				return false
+			}
 		},
```

Comment:
Kinda expected, but this means that if the `subpath` is a symlink, the only time a new commit would be registered is if the symlink itself changed (i.e. to a different path), and not when the underlying target dir changes.

I guess we can call it out in the release notes, but I feel like this is obvious enough that we don't need to worry about it

Reply:
Oh, it should also match on `strings.HasPrefix`, `strings.Contains` will match if any subdir has the same name (i.e. `internal` should only match `./internal`, not `./pkg/some-pkg/internal`).

You might also need to watch out for `internal` and `./internal`, which should logically match the same thing.